src/Groups/directproducts.jl

```diff
@@ -0,0 +1,486 @@
+Return the `j`-th factor of `G`.
+"""
+function factorofdirectproduct(G::DirectProductOfGroups, j::Base.Integer)
+   if j in 1:G.n return G.L[j]
+   else throw(ArgumentError("index not valid"))
+   end
+end
+
+"""
+    as_perm_group(G::DirectProductOfGroups)
+If `G` is direct product of permutations groups, return `G` as permutation group.
+"""
+function as_perm_group(G::DirectProductOfGroups)
+   if [typeof(H)==PermGroup for H in G.L]==[true for i in 1:G.n]
+      return PermGroup(G.X, GAP.Globals.Maximum(GAP.Globals.MovedPoints(G.X)))
+   else
+      throw(ArgumentError("The group is not a permutation group"))
+   end
+end
+
+"""
+    as_polycyclic_group(G::DirectProductOfGroups)
+If `G` is direct product of polycyclic groups, return `G` as polycyclic group.
+"""
+function as_polycyclic_group(G::DirectProductOfGroups)
+   if [typeof(H)==PcGroup for H in G.L]==[true for i in 1:G.n]
+      return PcGroup(G.X)
+   else
+      throw(ArgumentError("The group is not a polycyclic group"))
+   end
+end
+
+"""
+    as_matrix_group(G::DirectProductOfGroups)
+If `G` is direct product of matrix groups over the ring `R` of dimension `n_1`, ... , `n_k` respectively, return `G` as matrix group over the ring `R` of dimension `n_1 + ... + n_k`.
+"""
+function as_matrix_group(G::DirectProductOfGroups)
+#   if S==MatrixGroup && T==MatrixGroup && GAP.Globals.FieldOfMatrixGroup(G.G1.X)==GAP.Globals.FieldOfMatrixGroup(G.G2.X)
+   if [typeof(H)==MatrixGroup for H in G.L]==[true for i in 1:G.n] && length(Set([base_ring(H) for H in G.L]))==1
+      return MatrixGroup(G.X)
+   else
+      throw(ArgumentError("The group is not a matrix group"))
+   end
+end
+
+"""
+    embedding(G::DirectProductOfGroups, j::Integer)
+Return the embedding of the `j`-th component of `G` into `G`, for `j` = 1,...,#factors of `G`.
+"""
+function embedding(G::DirectProductOfGroups, j::Base.Integer)
+   if !G.isfull
+      throw(ArgumentError("Embedding is not defined for proper subgroups of direct products"))
+   end
+   if j in 1:G.n
+      f=GAP.Globals.Embedding(G.X,j)
+      gr=G.L[j]
+      typeG=typeof(gr)
+   else
+      throw(ArgumentError("index not valid"))
+   end
+   return GAPGroupHomomorphism{typeG,DirectProductOfGroups}(gr,G,f)
+end
+
+"""
+    projection(G::DirectProductOfGroups, j::Integer)
+Return the projection of `G` into the `j`-th component of `G`, for `j` = 1,...,#factors of `G`.
+"""
+function projection(G::DirectProductOfGroups, j::Base.Integer)
+  # @assert W.isfull "Projection not defined for proper subgroups of wreath products"
+   f=GAP.Globals.Projection(G.Xfull,j)
+   if j in 1:G.n
+      H = G.L[j]
+      if !G.isfull
+         Gf = DirectProductOfGroups(G.Xfull,G.L,G.n,G.Xfull,true)
+         g = embedding(Gf,G)
+         p = g*GAPGroupHomomorphism{DirectProductOfGroups,typeof(H)}(Gf,H,f)
+      else
+         p = GAPGroupHomomorphism{DirectProductOfGroups,typeof(H)}(G,H,f)
+      end
+      return hom(G,H,x->p(x))
```

Comment:
... and why isn't this:
```suggestion
      return hom(G,H,p)
```
For that matter: Why isn't it:
```suggestion
      return p
```
i.e., what exactly does `hom` do here?

Reply:
This is related to the previous question. Now I see that one of the operations useless, but at least one of them has to be maintained. So, we can either replace `hom(G,H,x->p(x))` by `p`, or replace the lines 131-137 by the single line `p = GAPGroupHomomorphism{DirectProductOfGroups,typeof(H)}(G,H,f)`.